walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,94 @@
+    List<TimeRange> busyTimesStart = new ArrayList<TimeRange>();
+    List<TimeRange> busyTimesEnd = new ArrayList<TimeRange>();
+    for (Event event : events) {
+      if (attendanceRequired(event, request)){
+        busyTimesStart.add(event.getWhen());
+        busyTimesEnd.add(event.getWhen());
+      }
+    }
+    Collections.sort(busyTimesStart, TimeRange.ORDER_BY_START);
+    Collections.sort(busyTimesEnd, TimeRange.ORDER_BY_END);
+
+    // Store a list of times when no attendees are busy
+    List<TimeRange> allAvailableTimes = new ArrayList<TimeRange>();
+    
+    // Return empty list if requested event is longer than the whole day
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) { return allAvailableTimes; }
+
+    // If no required attendees are busy, return the whole day for availability
+    int totalEvents = busyTimesStart.size();
+    if (totalEvents == 0) {
+      allAvailableTimes.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TimeRange.END_OF_DAY, true));
+      return allAvailableTimes;
+    }
+
+    // Check for free time before first scheduled event.
+    int first = 0;
+    if (busyTimesStart.get(first).start() > TimeRange.START_OF_DAY) {
+      allAvailableTimes.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, busyTimesStart.get(first).start(), false));
+    }
+
+    int curStartIndex = 0, curEndIndex = 1, lastEventIndex = totalEvents - 1;
+
+    while (curStartIndex < totalEvents && curEndIndex < totalEvents) {
+      while (curEndIndex < lastEventIndex && busyTimesStart.get(curStartIndex).overlaps(busyTimesEnd.get(curEndIndex))) {
+        ++curEndIndex;
+      }
+
+      if (curEndIndex == lastEventIndex) { 
+        if (busyTimesStart.get(curStartIndex).end() < busyTimesEnd.get(curEndIndex).start()) {
+          allAvailableTimes.add(TimeRange.fromStartEnd(busyTimesStart.get(curStartIndex).end(), busyTimesEnd.get(curEndIndex).start(), false));
+        }
+        break;   
+      }
+
+      allAvailableTimes.add(TimeRange.fromStartEnd(busyTimesStart.get(curStartIndex).end(), busyTimesEnd.get(curEndIndex).start(), false));
+      
+      ++curStartIndex;
+      while(curStartIndex < lastEventIndex && busyTimesStart.get(curStartIndex).start() < busyTimesEnd.get(curEndIndex).start()) {
+        ++curStartIndex;
+      }
+    }
+
+    // Check for free time after last scheduled event.
+    int last = totalEvents-1;
+    if (busyTimesEnd.get(last).end() < TimeRange.END_OF_DAY) {
+      int latest = Math.max(busyTimesStart.get(last).end(), busyTimesEnd.get(last).end());
+      allAvailableTimes.add(TimeRange.fromStartEnd(latest, TimeRange.END_OF_DAY, true));
+    }
+
+    // Only return available times that are long enough for the requested duration.
+    List<TimeRange> availableTimes = new ArrayList<TimeRange>();
+    for (TimeRange time : allAvailableTimes) {
+      if (request.getDuration() <= time.duration()) { availableTimes.add(time); }
+    }
```

Comment:
I like waiting to filter by this condition afterwards.

You can use Java 8 streams to filter this. Otherwise, rename one of ```availableTimes``` and ```allAvailableTimes``` to something that actually describes the different between the two.